### PR TITLE
OKTA-389639 [.NET] Handle local form field level errors from IDXResponse

### PR DESCRIPTION
**Pull request by andriizhegurov-okta**
Added support for Ion form validation error messages

Top-level error messages are already being extracted from responses like this:
```json {
  "version": "1.0.0",
  "stateHandle": "02lQUSR909HYCWWIlWKrmlvDuukREMzx0DrWZQwGIg",
  "expiresAt": "2021-04-09T19:19:18.000Z",
  "intent": "LOGIN",
  "remediation": {
"................................"
  },
  "messages": {
    "type": "array",
    "value": [
      {
        "message": "Authentication failed",
        "i18n": {
          "key": "errors.E0000004"
        },
        "class": "ERROR"
      }
    ]
  },
  "currentAuthenticator": {
".............................. "
}
}
```

In this PR I'm extracting form validation errors, like this one:
```json
{
  "version": "1.0.0",
  "stateHandle": "028e-r0fW99Ud4dADthU5wogZhdxzLhvOUPnC6FXVx",
  "expiresAt": "2021-04-27T22:05:50.000Z",
  "intent": "LOGIN",
  "remediation": {
    "type": "array",
    "value": [
      {
        "rel": [
          "create-form"
        ],
        "name": "reset-authenticator",
        "relatesTo": [
          "$.currentAuthenticator"
        ],
        "href": "..................",
        "method": "POST",
        "produces": "application/ion+json; okta-version=1.0.0",
        "value": [
          {
            "name": "credentials",
            "type": "object",
            "form": {
              "value": [
                {
                  "name": "passcode",
                  "label": "New password",
                  "secret": true,
                  "messages": {
                    "type": "array",
                    "value": [
                      {
                        "message": "Password requirements were not met. Password requirements: at least 8 characters, a lowercase letter, an uppercase letter, a number, no parts of your username. Your password cannot be any of your last 4 passwords.",
                        "i18n": {
                          "key": "password.passwordRequirementsNotMet",
                          "params": [
                            "Password requirements: at least 8 characters, a lowercase letter, an uppercase letter, a number, no parts of your username. Your password cannot be any of your last 4 passwords."
                          ]
                        },
                        "class": "ERROR"
                      }
                    ]
                  }
                }
              ]
            },
            "required": true
          },
          {
            "name": "stateHandle",
            "required": true,
            "value": "028e-r0fW99Ud4dADthU5wogZhdxzLhvOUPnC6FXVx",
            "visible": false,
            "mutable": false
          }
        ],
        "accepts": "application/json; okta-version=1.0.0"
      }
    ]
  },
  "currentAuthenticator": {
    "type": "object",
    "value": {
      "type": "password",
      "key": "okta_password",
      "id": "..................",
      "displayName": "Password",
      "methods": [
        {
          "type": "password"
        }
      ],
      "settings": {
        "complexity": {
          "minLength": 8,
          "minLowerCase": 1,
          "minUpperCase": 1,
          "minNumber": 1,
          "minSymbol": 0,
          "excludeUsername": true,
          "excludeAttributes": []
        },
        "age": {
          "minAgeMinutes": 0,
          "historyCount": 4
        }
      }
    }
  },
  "authenticators": {
    "type": "array",
    "value": [
      {
    ".................."
      }
    ]
  },
  "authenticatorEnrollments": {
    "type": "array",
    "value": [
      {
"    .................."
      },
      {
    ".................."
      }
    ]
  },
  "recoveryAuthenticator": {
   " .................."
  },
  "user": {
   " .................."
  },
  "cancel": {
    ".................."
  },
  "app": {
    ".................."
    }
  }
}
```

### CODE REVIEW

File: src/Okta.Sdk.Abstractions/IonApiError.cs [hunk 1]
```diff
@@ -38,5 +51,14 @@ private string GetErrorSummary()
 
             return sbErrorSumary.ToString();
         }
+
+        private string GetFormValidationErrorSummary()
+        {
+            var jToken = JToken.Parse(GetRaw());
+            var errorMessages = jToken
+                                    .SelectTokens("$.remediation.value[*].value[*].form.value[*].messages.value[*].message")
```
Comment (bryanapellanes-okta) [inline src/Okta.Sdk.Abstractions/IonApiError.cs:59]: What's the reason we're using this form of traversal (this isn't the first example I've seen of this)?  I find this hard to read and understand, but maybe there's a reason for doing it this way versus, modeling the remediation, for example.

**Review by bryanapellanes-okta — COMMENTED** [review]
Question...

File: src/Okta.Sdk.Abstractions/IonApiError.cs (hunk 1 above)
Comment (andriizhegurov-okta) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:59]: Yes, that's how it's currently done in other SDKs. To my understanding we are doing it this way because the response format is expected to change. So we will have less refactoring to do in future.

Comment (arvindkrishnakumar-okta) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:59]: While this is a simple one liner that works, it is quite hard to follow.

Comment (laura-rodriguez) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:59]: So keeping in mind we didn't want to model this particular response, what do you suggest @arvindkrishnakumar-okta? I'm curious to know if you guys have any other ideas.

Comment (arvindkrishnakumar-okta) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:59]: Didn't get to this ticket yet on the Java SDK side. We need to explore if Java has a similar framework that can grep for such form level errors inside a Json with a regex. If not, we'd have to tediously loop through each and every form and look for the values in it to see if there are any errors and then collect them.

cc: @sergiishamrai-okta

Comment (laura-rodriguez) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:172]: Right, but if you had both approaches available in java, what would you use and why? @arvindkrishnakumar-okta

Comment (bretterer) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:172]: I see no issues with this method right now. In the future, we will hopefully have an top level messages object we can use, but for now, this works

Comment (arvindkrishnakumar-okta) [reply src/Okta.Sdk.Abstractions/IonApiError.cs:172]: I am fine with this approach that works at the moment (assuming we'll have root level messages soon).

**Review by laura-rodriguez — APPROVED** [review]
🚢
